src/ext/machnet_common.h

```diff
@@ -539,6 +552,63 @@ __machnet_channel_buf_free_bulk(const MachnetChannelCtx_t *ctx, uint32_t n,
   // multi-consumer function.
   return jring_mp_enqueue_bulk(buf_ring, bufs, n, NULL);
 }
+static inline __attribute__((always_inline)) uint32_t
+__machnet_channel_buf_free_cached(MachnetChannelCtx_t *ctx, uint32_t cnt,
+                                  MachnetRingSlot_t *buffer_indices) {
+  uint32_t ret = 0;
+
+  if (ctx->cached_buf_available == 0) {
+    // If cache is empty, fill the cache directly
+    for (ret = 0; ret < cnt; ret++) {
+      ctx->cached_buf_indices[ret] = buffer_indices[ret];
+      ctx->cached_buf_available++;
+    }
+    ctx->cached_buf_index = 0;
+  } else {
+    // Determine start index and step based on cached buffer's index
+    uint32_t start = (ctx->cached_buf_index == 0) ? ctx->cached_buf_available
+                                                  : ctx->cached_buf_index - 1;
+    int step = (ctx->cached_buf_index == 0) ? 1 : -1;
+
+    // Cache is partially filled, so fill it accordingly
+    // If pointer is at the beginning, fill forwards, otherwise fill backwards
+    for (uint32_t i = start; ret < cnt && ctx->cached_buf_index;
+         ret++, i += step) {
+      ctx->cached_buf_indices[i] = buffer_indices[ret];
+      ctx->cached_buf_available++;
+      if (step == -1) ctx->cached_buf_index--;
+    }
+
+    // Fill any remaining space at the end of the cache
+    uint32_t endPaddingIndex =
+        ctx->cached_buf_index + ctx->cached_buf_available;
+    for (uint32_t i = endPaddingIndex; ret < cnt; ret++, i++) {
+      ctx->cached_buf_indices[i] = buffer_indices[ret];
+      ctx->cached_buf_available++;
+    }
+  }
+
+  // Ensure that we processed all buffer indices
+  assert(ret == cnt);
+  return ret;
+}
+
+static inline __attribute__((always_inline)) uint32_t
+__machnet_channel_buf_free(MachnetChannelCtx_t *ctx, uint32_t cnt,
+                           MachnetRingSlot_t *buffer_indices) {
+  uint32_t available_capacity, to_free, ret;
+  available_capacity = CACHED_BUF_SIZE - ctx->cached_buf_available;
+  to_free = MIN(cnt, available_capacity);
+  ret = (to_free)
+            ? __machnet_channel_buf_free_cached(ctx, to_free, buffer_indices)
```

Comment:
I might be confused or missed something (this is a quick pass), but how do you know who is the caller of this application? On the master version all buffer-related routines are used by both sides, and then that'd be not MP-safe?

Reply:
Thank you, @marinosi,

To clarify, the functions `__machnet_channel_buf_free_cached` and `__machnet_channel_buf_free` are indeed used exclusively within the `machnet_recvmsg` function, which is intended for use on the application side.

The `__machnet_channel_buf_free_bulk` function is designed to be MP-safe, as it relies on `__jring_mp_enqueue_bulk` for its operations.

In essence, `__machnet_channel_buf_free` acts as a simple wrapper to avoid duplicating the decision-making code for freeing buffers to either the cache or `buf_ring`.

Please LMK if you are not convinced by the answer.

Reply:
I see; I originally wrote the code in `machnet_commmon` to be used by any side unless there is indicated in the helper function name. Could you please move this implementation in each own helper functions in machnet.c to isolate the app-side implementation there as we did for the channel? You already handling the bulk allocations accordingly in `sendmsg`.

Reply:
Done